onadata/apps/api/viewsets/tableau_viewset.py

```diff
@@ -0,0 +1,305 @@
+from rest_framework.response import Response
+
+from onadata.libs.data import parse_int
+from onadata.libs.renderers.renderers import pairing
+from onadata.apps.logger.models import Instance
+from onadata.apps.logger.models.xform import XForm
+from onadata.apps.api.viewsets.open_data_viewset import (
+    OpenDataViewSet)
+from onadata.libs.serializers.data_serializer import TableauDataSerializer
+from onadata.libs.utils.common_tags import (
+    ID, MULTIPLE_SELECT_TYPE, REPEAT_SELECT_TYPE, PARENT_TABLE, PARENT_ID)
+
+
+DEFAULT_TABLE_NAME = 'data'
+GPS_PARTS = ['latitude', 'longitude', 'altitude', 'precision']
+
+
+def process_tableau_data(
+        data, xform,
+        parent_table: str = None,
+        parent_id: int = None,
+        current_table: str = DEFAULT_TABLE_NAME):
+    result = []
+    if data:
+        for idx, row in enumerate(data, start=1):
+            flat_dict = defaultdict(list)
+            row_id = row.get('_id')
+
+            if not row_id and parent_id:
+                row_id = int(pairing(parent_id, idx))
+                flat_dict[PARENT_ID] = parent_id
+                flat_dict[PARENT_TABLE] = parent_table
+                flat_dict[ID] = row_id
+            else:
+                flat_dict[ID] = row_id
+
+            for (key, value) in row.items():
+                qstn = xform.get_element(key)
+                if qstn:
+                    qstn_type = qstn.get('type')
+                    qstn_name = qstn.get('name')
+
+                    prefix_parts = [
+                        question['name'] for question in qstn.get_lineage()
+                        if question['type'] == 'group'
+                    ]
+                    prefix = "_".join(prefix_parts)
+
+                    if qstn_type == REPEAT_SELECT_TYPE:
+                        repeat_data = process_tableau_data(
+                            value, xform,
+                            parent_table=current_table,
+                            parent_id=row_id,
+                            current_table=qstn_name)
+                        cleaned_data = unpack_repeat_data(
+                            repeat_data, flat_dict)
+                        flat_dict[qstn_name] = cleaned_data
+                    elif qstn_type == MULTIPLE_SELECT_TYPE:
+                        picked_choices = value.split(" ")
+                        choice_names = [
+                            question["name"] for question in qstn["children"]]
+                        list_name = qstn.get('list_name')
+                        unpack_select_multiple_data(
+                            picked_choices,
+                            list_name,
+                            choice_names,
+                            prefix,
+                            flat_dict)
+                    elif qstn_type == 'geopoint':
+                        gps_parts = unpack_gps_data(
+                            value, qstn_name, prefix)
+                        flat_dict.update(gps_parts)
```

Comment:
if `unpack_gps_data` returns `None`, line 79 would trigger an error, no?

Reply:
I would say no. I dont believe there could be a scenerio where it would return `None`. Maybe this is something new we could test

The function can only be called if the element is of "geopoint" question type. All geopoint question types are space separated strings, unless the user does not provide data for this question, for which the field is then not present in the json returned by the serializer. I could be wrong Cc: @DavisRayM